local-storage/src/block.rs

```diff
@@ -0,0 +1,263 @@
+use anyhow::{bail, Result};
+use bytes::Bytes;
+use cid::Cid;
+use iroh_unixfs::Block;
+use std::collections::HashSet;
+use std::str::FromStr;
+
+#[derive(Debug, PartialEq)]
+pub struct StoredBlock {
+    pub cid: String,
+    pub data: Vec<u8>,
+    pub links: Vec<String>,
+}
+
+impl StoredBlock {
+    pub fn validate(&self) -> Result<()> {
+        // For now we'll just piggy back on the validate logic built
+        // into n0/beetle:unixfs::Block
+        let block: Block = self.try_into()?;
+        block.validate()
+    }
+}
+
+impl TryInto<Block> for &StoredBlock {
+    type Error = anyhow::Error;
+
+    fn try_into(self) -> std::result::Result<Block, Self::Error> {
+        let cid = Cid::from_str(&self.cid)?;
+        let links = self
+            .links
+            .iter()
+            .map(|l| Cid::from_str(l).unwrap())
+            .collect::<Vec<Cid>>();
+        let data = Bytes::from(self.data.clone());
+        Ok(Block::new(cid, data, links))
+    }
+}
+
+pub fn validate_dag(stored_blocks: Vec<StoredBlock>) -> Result<()> {
+    if stored_blocks.is_empty() {
+        bail!("No blocks found in dag")
+    }
+    let blocks = stored_blocks
+        .iter()
+        .map(|b| b.try_into().unwrap())
+        .collect::<Vec<Block>>();
+    // First check if all blocks are valid
+    for block in blocks.iter() {
+        block.validate()?;
+    }
+    // If there is only one block, and it is valid, then assume it's valid on its own
+    if blocks.len() == 1 {
```

Comment:
not sure that we can assume a single block is valid. We should ensure it is also a root block, i.e. nothing points to it. It should be safe to remove this check altogether.